homeassistant/components/synology_dsm/__init__.py

```diff
@@ -221,6 +233,46 @@ async def _async_update_listener(hass: HomeAssistantType, entry: ConfigEntry):
     await hass.config_entries.async_reload(entry.entry_id)
 
 
+async def _async_setup_services(hass: HomeAssistantType):
+    """Service handler setup."""
+
+    async def service_handler(call: ServiceCall):
+        """Handle service call."""
+        _LOGGER.debug(
+            "service_handler - called as '%s' with data: %s", call.service, call.data
+        )
+        serial = call.data.get(CONF_SERIAL)
+        dsm_devices = hass.data[DOMAIN]
+
+        if serial:
+            dsm_device = dsm_devices.get(serial)
+        elif len(dsm_devices) == 1:
+            dsm_device = next(iter(dsm_devices.values()))
+            serial = next(iter(dsm_devices))
+        else:
+            _LOGGER.error(
+                "service_handler - more than one DSM configured, must specify one of serials %s",
+                sorted(dsm_devices),
+            )
+            return
+
+        if not dsm_device:
+            _LOGGER.error(
+                "service_handler - DSM with specified serial %s not found", serial
+            )
+            return
+
+        _LOGGER.info("%s DSM with serial %s", call.service, serial)
+        dsm_api = dsm_device[SYNO_API]
+        if call.service == SERVICE_REBOOT:
+            await dsm_api.async_reboot()
+        elif call.service == SERVICE_SHUTDOWN:
+            await dsm_api.system.shutdown()
+
+    for service in SERVICES:
+        hass.services.async_register(DOMAIN, service, service_handler)
```

Comment:
just to be sure: register the same service multiple times causes no issue ?

(when multiple NAS)

Reply:
during tests with two configured DSM:

- each service still only registered once
- when service is triggered it is also only processed once

so I would assume there are no problems 😃